mlton/backend/ssa2-to-rssa.fun

```diff
@@ -1642,38 +1642,237 @@ fun convert (program as S.Program.T {functions, globals, main, ...},
+               case var of
+                  NONE => pushKeep st
+                | SOME var =>
+                  let
+                     datatype location = datatype Static.Location.t
+                     fun keep () = pushKeep st
+                     fun keepWithStatic'' s =
+                        (setGlobalStatic (var, s)
+                         ; keep ())
+                     fun copy var' = keepWithStatic'' (globalStatic var')
+                     fun keepWithStatic' s = keepWithStatic'' (SOME s)
+                     fun keepWithStatic s = keepWithStatic' (SOME s)
+                     fun word w = keepWithStatic (Word w)
+                     fun real r = keepWithStatic (Real r)
+                     fun static (soe: PackedRepresentation.StaticOrElem.t) =
+                        case toRtype ty of
+                           SOME rty =>
+                              (case soe of
+                                  Static s =>
+                                     (pushStatic (var, ty, rty, s);
+                                      case Static.location s of
+                                         Heap => () (* address of heap static not a valid elem *)
+                                       | _ => setGlobalStatic (var, SOME (SOME (Address var))))
+                                | Elem e => keepWithStatic e)
+                         | _ => keepWithStatic' NONE
+                  in
+                     case exp of
+                        S.Exp.Const c =>
+                           (case c of
+                               Const.WordVector wxv =>
+                                  if !Control.staticAllocWordVectorConsts
+                                     then static (makeWordXVector {data = wxv,
+                                                                   location = Static.Location.ImmStatic,
+                                                                   ty = ty})
+                                     else keep ()
+                             | Const.Word w => word w
+                             | Const.Real r => real r
+                             | _ => keep ())
+                      | S.Exp.Inject {variant, ...} =>
+                           copy variant
+                      | S.Exp.Object {args, con} =>
+                           (case !Control.staticInitObjects of
+                               NONE => keep ()
+                             | SOME staticInitHeapObjects =>
+                                  if Vector.exists (args, Option.isNone o globalStatic)
+                                     then keep ()
+                                     else let
+                                             val location = getLocation (ty, !Control.staticAllocObjects, false)
+                                          in
+                                             if location <> Static.Location.Heap
+                                                orelse staticInitHeapObjects
```

Comment:
I was struggling to find/write a source program that, with `-static-alloc-internal-ptrs static -static-alloc-objects true -static-init-objects all`, would trigger this code path that leads to a statically initialized but dynamically allocated object.  Eventually, I realized that there should be no such objects.  With `-static-alloc-internal-ptrs static -static-alloc-objects true`, the only objects that will be statics with a `Heap` location are objects with a mutable `Objptr` field.  Moreover, only global objects are eligible to be statics.  But, assuming a safe-for-space globalization pass, there should be no global objects with mutable `Objptr` fields (because global objects are roots for the whole program execution).  I seem to remember @jasoncarr0 making some comment that was supposed to suggest this (something along the lines of "by fiat, should never happen"), but I can't find it right now.

Using `-globalize-small-types 9`, which allows the globalization of arbitrary `array` and `ref` objects (irrespective of the "size" of their contents), then we can trigger global objects with mutable `Objptr` fields that become statically initialized, dynamically allocated.

With this observation about safe-for-space, I'm not understanding why there should ever be a non-empty `globalObjptr[]` array.  In a self-compile, there are 26.

Reply:
It can happen, if `globalize-small-type 4` is on, then an example would be a mutable field of a tuple of total size more than 64 bytes. It is safe to globalize, but will turn into an Objptr in Rssa. With the current code, intInfs aren't included only because they're infrequent, so that would explain some globalObjptrs on a self-compile.

Reply:
Unless I've made a mistake in understanding packed-representation here

Reply:
You are right, that a `(int * int) ref` could be space-safely globalized and could result in an object with a mutable `Objptr` field.  However, it is also likely that such a tuple would be `RefFlatten`ed; I suspect that is why we don't see many of them.  As you say with `-globalize-small-types 4`, then we might have a `(int * int, int * int) either ref` be globalized.

Also, it turns out that we "grandfathered" `IntInf.int` as a "small type"; so a `IntInf.int ref` can be globalized, but is represented as an object with a mutable `Objptr` field.

Reply:
Finally, we might have an `IntInf.int ref ref`; the inner ref might be statically initialized (but dynamically allocated), so the outer ref could not be statically initialized.